### DISCUSSION TITLE
Getting arround "rate-limited" errors? (RealTime / broadcast / subscribe / track)

### DISCUSSION THREAD

**Discussion by jilpi** [supabase/supabase#12575] [Questions]
Hi everyone,

We are quite new to Supabase. Our application is a WebRTC-based video conferencing system.

We use two functions of Realtime:
 - broadcasting messages (as part of a WebRTC negotiation - sending offers and ICE candidates to the other peer)
 - and tracking "presence states" (to know who currently joins the room)

When handling WebRTC ICE Candidate gathering, most candidates are gathered very fast. Usually 10-20 candidates in a very short period of time (few 10s of milliseconds). If we send them all in "real time", as they are gathered, it immediately fails with client-generated "rate-limited" errors.
The way we have circumvented this is by introducing a "retry" mechanism.

```js
   const RT_DEFAULT_RETRY_COUNT = 10 // retry sending max 10 times

    public sendBroadcastMessage = async (
        message: Message,
        ttl: number = RT_DEFAULT_RETRY_COUNT
    ): Promise<void> => {

        const response = await this.channel?.send({
            type: 'broadcast',
            event: "message",
            payload: message,
        })

        if (response !== "ok") {
            if (ttl <= 0) {
                console.error("Message could not be sent through RealTimeChannel, TTL <= 0", message);
                return;
            }
            // Trying to send again, with a delay to avoid flooding the server
            const delay = this.rtRandomInterval(); // function returning a random integer between an app-level defined interval
            setTimeout(this.sendBroadcastMessage.bind(this), delay, message, ttl - 1);
        }
    }
```

We certainly should have done something more complicated (like setting a stack of messages, and sending them one by one at a given rate). But, well, our quick and dirty approach works.

Now, we realise that we are getting this kind of 'rate-limited' errors even when we subscribe and begin to "track" our presence.

```js
        this.channel
            .on('presence', { event: 'sync' }, () => this.onPresenceChanged())
            .on('presence', { event: 'join' }, ({ key, newPresences }) => { console.debug("New clients joined channel: ", key, newPresences) })
            .on('presence', { event: 'leave' }, ({ key, leftPresences }) => { console.debug("Clients left channel: ", key, leftPresences) })
            .on("broadcast",
                { event: "message" },
                (event) => {
                    this.onBroadcastMessage(event.payload); // Here we handle whatever message we received
                }
            )
            .subscribe(async (status, err) => {
                console.debug(`Channel subscription status: ${status}`)
                if (status === 'SUBSCRIBED') {
                    const presenceTrackStatus = await this.channel?.track({
                        userId: this.userId,
                        userName: this.userName,
                        userKind: this.userKind,
                        online_at: new Date().toISOString(),
                    })
                    console.debug('presenceTrackStatus = ' + presenceTrackStatus); // VERY FREQUENT RATE-LIMITED HERE
                } else {
                    console.error(`Channel subscription failed with error: `, err)
                };
            })
```

This code is very close from the examples found in [Supabase documentation](https://supabase.com/docs/guides/realtime/presence#track-presence). So getting errors here is very surprising.

We start to wonder if we are doing something wrong, and would appreciate it if you could share your thoughts.

Thanks

**Comment by bobmccracken** [comment]
It's been a while, so I'm guessing the OP probably worked this out or switched off of Supabase for realtime (probably not a bad idea). Regardless, I'm leaving my notes here.

tl;dr increase your client side rate limit and always `await` your `send`s and `track`s

You can adjust the client side rate limit when creating the client with `createClient`. See https://supabase.com/docs/guides/realtime/broadcast#client-side-rate-limit. The docs are a _little_ confusing here because the setting is `eventsPerSecond`, but the docs suggest this actually translates into 1 message per X milliseconds. So, if `eventsPerSecond` is the default 10, you get 1 message per 100 milliseconds.

I'm still not 100% sure how the rate limiting is actually implemented, but my experience leads me to believe that the client is actually checking that you don't exceed 1 message per X milliseconds. That experience I'm referencing is also important.

I'm building an app that uses both presence and broadcast. Fwiw, both of these seem to get put in the same rate limit bucket. The important thing here is I was firing off `send`s and `track`s without always `await`ing them. Turns out this can be pretty important if those calls happen in close proximity. I was getting rate limited after firing off only 2 or 3 events, which really had me baffled for a while. 2 or 3 seemed well below the client limit, especially after I increased it to 1000 events per second.

After closely reading the docs, I decided maybe this ms based limiting was the issue because the events I was triggering were happening almost simultaneously. So, I put an await on everything and I stopped getting limited. My working theory now is that the await allowed just enough time between events to avoid the rate limit, which is ms based.

I was gonna say I can't really be certain about the rate limiting method without looking at the code, so I looked at the code. This seems to confirm it as being ms based https://github.com/supabase/realtime-js/blob/master/src/RealtimeClient.ts#L102.

Browsing over there also lead me to an [Issue](https://github.com/supabase/realtime-js/issues/205) that touches upon everything in my post here. Wish I had read that before I lost so much hair.

Oh I also discovered (in the README) that you can turn off the client rate limiting entirely with `-1` if you want, but then you have to update the "server's rate limiting", whatever that means. Maybe just the plan quotas? I dunno. Docs around realtime are pretty shite. Almost certain I'll be switching to some other realtime platform after finishing this POC.

**Comment by jilpi** [comment]
> I'm guessing the OP probably worked this out

Not really :-(
We made something barely workable by overusing the retry mechanism and using other gimmicks. Very dirty, and we don't like it at all, even for a POC. We don't have confidence to move to production with the current solution.

> Oh I also discovered (in the README) that you can turn off the client rate limiting entirely with -1 

Nice find, thanks for sharing :-)

It's true that the documentation is really lacking. A pity because if known limitations were documented, it would save a lot of time and frustration.
